src/pages/index.js

```diff
@@ -15,86 +16,215 @@ import {
+        card._isLiked = true;
+      })
+      .catch((err) => {
+        console.error(`Error on Card Like ${err}`);
+      });
+  }
+}
+
+//add card functionality
 const createCard = (data) => {
-  const card = new Card(data, "#card-template", handleImagePreview);
+  const card = new Card(
+    data,
+    "#card-template",
+    handleImagePreview,
+    handleDeleteCard,
+    handleLikeClick
+  );
   return card.getCardElement();
 };
 
-//render card elements
-const cardListEl = new Section(
-  {
-    items: initialCards,
-    renderer: (data) => {
-      cardListEl.addItem(createCard(data));
-    },
-  },
-
-  ".cards__list"
+const addCardPopup = new PopupWithForm(
+  "#add-card-modal",
+  handleAddCardFormSubmit
 );
+addCardPopup.setEventListeners();
+
+addNewCardButton.addEventListener("click", () => {
+  addCardPopup.open();
+});
 
-cardListEl.renderItems();
+function handleAddCardFormSubmit(data) {
+  const name = data.title.trim();
+  const link = data.link.trim();
+  addCardPopup.close();
+  addCardFormElement.reset();
+  addCardValidator.disableSubmitButton();
+  api
+    .createCard({ name, link })
+    .then((res) => {
+      const card = createCard(res);
+      section.addItem(card);
+    })
+    .finally(() => {
+      addCardPopup.setLoading(false);
+    });
+}
 
-//preview Image Modal
+//preview card image functionality
 const previewImageModal = new PopupWithImage("#preview-image-modal");
 previewImageModal.setEventListeners();
 
 function handleImagePreview(card) {
   previewImageModal.open(card);
 }
 
-//edit profile functionality
+//delete card functionality
+function handleDeleteCard(cardId, cardElement) {
+  deleteModalConfirmation.open(cardId, cardElement);
+  console.log(cardId, cardElement);
+}
+
+const deleteModalConfirmation = new PopupConfirmation(
+  "#modal-confirm",
+  (cardId, cardElement) => {
+    return api.deleteCard(cardId).then(() => {
+      cardElement.remove();
+    });
+  }
+);
+
+//USER INFO FUNCTIONS
+
+//get user info
 const userInfo = new UserInfo({
   name: ".profile__title",
   description: ".profile__description",
+  avatarSelector: ".profile__image",
 });
 
+api
+  .getUserInfo()
+  .then((userData) => {
+    userInfo.setUserInfo({
+      name: userData.name,
+      description: userData.about,
+      avatar: userData.avatar,
+    });
+  })
+  .catch((err) => console.error(err));
+
+//edit profile functionality
 const editProfilePopup = new PopupWithForm(
   "#profile-edit-modal",
   handleProfileEditSubmit
 );
 editProfilePopup.setEventListeners();
 
+function handleProfileEditSubmit(data) {
+  api
+    .editUserInfo({
+      name: data.title,
+      about: data.description,
+    })
+    .then((userData) => {
+      userInfo.setUserInfo({
+        name: userData.name,
+        description: userData.about,
+        avatar: userData.avatar,
+      });
+    })
+    .catch((err) => console.error(`Failed to update users info: ${err}`))
+    .finally(() => {
+      editProfilePopup.setLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```